0010-zkvm-keccak.md

```diff
@@ -0,0 +1,788 @@
+
+### Constraints
+
+
+> For efficiency reasons, the state is assumed to be stored in expanded form, so that back-and-forth conversions do not need to take place repeatedly.
+
+The support for the Keccak hash function will require the following gate types:
+
+**PRECONDITIONS:** the message to be hashed should be padded with the $10^*1$ rule before 
+
+- `KeccakSponge`: performs formatting actions related to the Keccak sponge.
+- `KeccakRound`: performs one full round of the Keccak permutation function.
+
+The high-level layout of the gates follows:
+
+| `KeccakSponge` | [0...100) | [100...168) | [168...200) | [200...400) | [400...800) |
+| -------------- | --------- | ----------- | ----------- | ----------- | ----------- |
+| Curr           | old_state | new_block   | zeros       | bytes      | shifts      |
+| Next           | xor_state |
+
+| `KeccakRound` | [0...440)   | [440...1540) | [1540...2440) | 
+| ------------- | ----------- | ------------ | ------------- | 
+| Curr          | theta_step  | pirho_step   |  chi_step     |
+
+| `KeccakRound` | [0...100)   |
+| ------------- | ----------- |
+| Next          | iota_step   |
+
+#### Sponge
+
+Support for padding shall be provided. In the Keccak PoC, this step took place at the Snarky layer. It checks that the correct amount of bits in the $10^*1$ rule are added until reaching a multiple of 1088 bits, and then adds 512 more zero bits to each block to form a full state.
+
+If the input was not previously expanded, the next step would be to expand all 25 words. Each word of 64 bits would have to be split into 4 parts of 16 real bits each. The expansion itself would be performed through the `Reset` lookup table. This step would require $4\times25=100$ lookups.
+
+There are two main modes: absorb and squeeze. Inside the absorb mode, there are two additional submodes called pad and root, which can as well happen at the same time. 
+
+
+- **Absorb mode**
+
+    In absorb mode, the first coefficient is set to 1: `absorb := coeff[0] = 1`, and the second coefficient is 0.
+
+    The gate takes a block as 68 chunks of 16 bits expanded, pads with zeros until reaching 100 chunks (1600 bits) to fit one whole state, and XORs it with the previous state.
+
+    The zero padding is enforced performing the following 32 constraints:
+
+    ```rust
+    constrain(absorb * zeros[i])
+    ```
+
+    The XOR part of the gate uses $100$ constraints 
+
+    ```rust
+    constrain(absorb * (xor_state[i] - (old_state[i] + new_state[i])))
```

Comment:
note (for me): and we have $xor_state[i] \in \mathbb{F}$

but it is in fact encoding a byte, i.e. the field value is below 2*16.

Reply:
Each i-th position of those constraints encode 2 bytes (16 bits) as 64 bit field elements.